### PR TITLE
BUG: fix out-of-bound evaluations in optimize.minimize, powell method

### PR DESCRIPTION
**Pull request by TianyiQ**
#### Reference issue
None

#### What does this implement/fix?
In [SciPy online document](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html) it is guaranteed that:

> (for optimize.minimize(mathod="powell")) if bounds are provided and the initial guess is within the bounds, then every function evaluation throughout the minimization procedure will be within the bounds. 

However the actual implemention did not follow this requirement, and this pull request fixes it.

### CODE REVIEW

**Comment by mdhaber** [comment]
Thanks @TianyiQ. I see that this will prevent this part of the code from evaluating the function at a point outside the bounds, but it's unclear what will happen afterwards. 

* What is the "construct the extrapolated point" part of the code doing? 
* If we `continue` to the next iteration here, is there a chance that the algorithm will get stuck (that didn't exist before)?
* Instead of `continue`ing, why not "construct the extrapolated point" with the exception that coordinates beyond the bound are replaced by the corresponding cooridinates of the bound?

On the other hand, the claim that "every function evaluation throughout the minimization procedure will be within the bounds" was added in gh-12172 (which has its roots in gh-10648), but it's unclear whether that was ever true. The statement was probably based on the fact that the _line search_ would not evaluate the function out of bounds, but this issue isn't occurring in the line search part at all. The other option is to just change the documentation, unless a very clearly "correct" fix can be found. Having bounds is still useful because we still guarantee that the _final solution_ will be within the bounds.

**Review by AtsushiSakai — COMMENTED** [review]
`_linesearch_powell` function updates `x` and `fval`, and this function already handle bounds constraint.
SO, I'm not sure why this code should be added....
```
if bounds and (np.any(lower_bound > x2) or np.any(x2 > upper_bound)):
            continue
```
Could you please elaborate on why this code should be added? @TianyiQ

**Comment by mdhaber** [comment]
@AtsushiSakai The documentation of Powell states that the function will never be evaluated at a point outside the bounds, but that's not true. You can try the test locally; the assertion will fail. The bounds should keep ` x > 0` but `func` is evaluated with `x` nearly `-0.5`.

This PR is an attempt to fix that. The code in the PR is probably not what we want, though. I'll wait for the author's reply to my questions, but it looks like an approach that works for the tested problem but may cause the algorithm to fail to converge to the solution for other problems.

I think we either want a solution that works for all (or "almost all", in some mathematical sense), or we should change the documentation's claims: the final solution will satisfy the bounds, but not all iterates will.

File: scipy/optimize/tests/test_optimize.py [hunk 1]
```diff
@@ -2028,6 +2028,15 @@ def fun(x):
 
     optimize.minimize(fun, x0=[0.6, 20], method='Powell', bounds=bounds)
 
+    # Another test from the original report - gh-13411
+    bounds = optimize.Bounds(lb=[0,], ub=[1,], keep_feasible=[True,])
+
+    def func(x):
+        assert x >= 0  and x <= 1
```
Comment (mdhaber) [inline scipy/optimize/tests/test_optimize.py:2035]:
```suggestion
        assert x >= 0 and x <= 1
```

**Comment by mdhaber** [comment]
Main Windows Python38-32bit-fast is taking unusually long, and I don't think it's caused by this PR. Hopefully it's not happening in other PRs, but in case it is, I wanted to note that it's happening here, too.

**Pull request merged**